llvm/lib/Transforms/Scalar/ConstraintElimination.cpp

```diff
@@ -1031,6 +1031,23 @@ void State::addInfoForInductions(BasicBlock &BB) {
   WorkList.push_back(FactOrCheck::getConditionFact(
       DTN, CmpInst::ICMP_SLT, PN, B,
       ConditionTy(CmpInst::ICMP_SLE, StartValue, B)));
+
+  assert(!StepOffset.isNegative() && "induction must be increasing");
+  // Try to add condition from header to the unique exit block, if there is one.
+  // When exiting either with EQ or NE, we know that the induction value must be
+  // u<= B, as a different exit may exit earlier.
+  if (Pred == CmpInst::ICMP_EQ) {
```

Comment:
It is incorrect when the initial value of indvar is greater than the bound.
Alive2: https://alive2.llvm.org/ce/z/UYsbAQ

We need `ConditionTy(CmpInst::ICMP_ULE, StartValue, B)` here.

Reply:
Great point thanks! Incorrectly assume that we only reach here for monotonically increasing cases, should have added tests... Done in https://github.com/llvm/llvm-project/commit/b7b8d028962e7a7280e3d65019345f2446785e7e

and update the code here